component/component.go

```diff
@@ -175,3 +175,10 @@ type CreateDefaultConfigFunc func() Config
 func (f CreateDefaultConfigFunc) CreateDefaultConfig() Config {
 	return f()
 }
+
+// GlobalID uniquely identifies a component
+type GlobalID struct {
+	ID         ID
+	Kind       Kind
+	PipelineID ID // Not empty only if the Kind is Processor
+}
```

Comment:
There are a couple nuances to the way we instance components within the service, which I think likely need to be reconciled in this PR.

Receivers and exporters are instanced _per data type_. For example, users will often configure a single `otlpexporter` for exporting multiple types of data, but the underlying implementation creates separate instances. which would share the same GlobalID.

This is probably the correct model to present externally, because it aligns with the user's understanding, but we will have to define how to reconcile with the underlying implementation. Minimally, instances of the "same" component must be able to safely publish updates concurrently.

This raises a usability complication though, because we may expect status thrashing or burying of problems. e.g. the traces instance of a component reports a problem, but the metrics instance immediately clobbers the status by reporting that it is healthy.

I think we probably need to maintain a separate statuses for each instance of the "same" component, and aggregate them whenever an event is published. e.g. report "lowest" health state of any instance that shares the GlobalID, regardless of which instance sent the most recent event.

To further complicate things, we have the `sharedcomponent` package, which allows component authors to opt in to a unified model. In this case, the service still reasons about separate instances, but there is in fact only one instance after all. For example, the `otlpreceiver` uses this mechanism so that the shared instance can present a single port. I'm not certain how best to handle this, but perhaps the `sharedcomponent` package needs to manage status reporting for the components that use it.

Connectors share the same concerns as receivers and exporters, except instancing is _per ordered pair of data types_. `sharedcomponent` may or may not be used here as well.

Reply:
🤯 this sounds somewhat complicated, but thanks for the explanation.

I'm wondering if we should assume that each instance of a component should report status individually, and if so, perhaps we can rename `GlobalID` to `InstanceID`, where the "ID" would have enough metadata to identify the underlying component. We could add some additional fields to the struct if needed.

I'm also wondering if this is accidentally already happening with the `GlobalID` as it's currently implemented. We are passing around pointers to GlobalID instances, via the host wrapper. Is the code that generates the GlobalIDs in graph generating a global ID per component instance? If so, we would have a GlobalID instance per component, each with a distinct pointer, but possibly the same struct values.

I'd also be interested in any other reasonable ways to approach this.

Reply:
> I'm also wondering if this is accidentally already happening with the GlobalID as it's currently implemented. We are passing around pointers to GlobalID instances, via the host wrapper. Is the code that generates the GlobalIDs in graph generating a GlobalID per component instance? If so, we would have a GlobalID instance per component, each with a distinct pointer, but possibly the same struct values.

To clarify, are you talking about [`graph.nodeID`](https://github.com/open-telemetry/opentelemetry-collector/blob/8aeb3ab04e695b626ae642c8c06f1d909fc33624/service/internal/graph/nodes.go#L37-L47) as the current implementation of `GlobalID`, and `service.Host`'s `graph.Graph` as the mechanism by which we're passing around pointers to the IDs?

Assuming so, I don't believe there are uniqueness problems there because the `nodeID`'s are hashes that are generated based on the the considerations I mentioned above. [Processor IDs](https://github.com/open-telemetry/opentelemetry-collector/blob/8aeb3ab04e695b626ae642c8c06f1d909fc33624/service/internal/graph/nodes.go#L120) are hashed with a specific pipeline ID. [Receiver IDs](https://github.com/open-telemetry/opentelemetry-collector/blob/8aeb3ab04e695b626ae642c8c06f1d909fc33624/service/internal/graph/nodes.go#L64) and [Exporter IDs](https://github.com/open-telemetry/opentelemetry-collector/blob/8aeb3ab04e695b626ae642c8c06f1d909fc33624/service/internal/graph/nodes.go#L168) are hashed with a data type, and [Connector IDs](https://github.com/open-telemetry/opentelemetry-collector/blob/8aeb3ab04e695b626ae642c8c06f1d909fc33624/service/internal/graph/nodes.go#L218) are hashed with an ordered pair of data types. `nodeID` is also not accessible outside of the graph package, and I don't think there's any way to indirectly reference or modify them.

> I'm wondering if we should assume that each instance of a component should report status individually, and if so, perhaps we can rename GlobalID to InstanceID, where the ID would have enough metadata to identify the underlying component. We could add some additional fields to the struct if needed.

I think it would be very hard for consumers of the events to make sense of this. The instancing is really an internal concern of the collector and not how we expect users to reason about the components.

> I'd also be interested in any other reasonable ways to approach this.

The only way I can think to do this is by modeling _component_ and _instance_ separately:
1. A _component_ model is presented to users, where each component is 1:1 with those defined in the configuration, except for processors which are understood to be unique to each pipeline.
2. An _instance_ model underlies the component model but is not directly exposed to users.
3. Every instance belongs to a component. i.e. A component may have one or more instances.
4. Internally, each instance reports its own status.
5. Externally, each component reports an aggregation of the statuses of its instances. (Often, a component will have only 1 instance, so its status is that of the instance)

Based on this, I think the host's `ReportComponentStatus` should effectively be `ReportInstanceStatus` and the reported status should be logged into a data structure which:
1. Maintains status for all instances
2. Understands the relationship between components and instances
3. Roughly, `map[GlobalComponentID]map[GlobalInstanceID]Status`

After updating the instance status, the `ReportInstanceStatus` function should aggregate the statuses for the relevant component and emit an event to StatusWatchers.

There's some plumbing to figure out, but what do you think about this model conceptually?

Reply:
> To clarify, are you talking about [graph.nodeID](https://github.com/open-telemetry/opentelemetry-collector/blob/8aeb3ab04e695b626ae642c8c06f1d909fc33624/service/internal/graph/nodes.go#L37-L47) as the current implementation of GlobalID, and service.Host's graph.Graph as the mechanism by which we're passing around pointers to the IDs?

I made some changes to graph in this PR. I ended up making a `map[graph.nodeID]*component.GlobalID` and due to idiosyncrasies of implementation, the component.GlobalID pointer is effectively an instance id.  That's not super important though.

> There's some plumbing to figure out, but what do you think about this model conceptually?

I'm ultimately open to any and all options, but there is one thing I wanted to bring up that likely complicates the ability to aggregate status at the component level. Right now we only have `StatusOK` and `StatusError`, but I would like a more expressive set of statuses for components to choose from. I was looking at [GRPC status](https://grpc.github.io/grpc/core/md_doc_statuscodes.html) as a possibility. I think there is a status for many, if not all issues a component could encounter. The 16 GRPC status codes are well-defined and carry a lot of information about the health of a component as a simple `code` and `message`. I'm not sure how you would aggregate multiple statuses from a system like GRPC into a single code though.

I wanted to do a little more homework on this idea before bringing it up, but now that I've mentioned it, what are your thoughts about adopting GRPC status for component status reporting? Would this change your thoughts about reporting status per instance vs status per component?

Reply:
If we did not want to go with GRPC status, we could come up with our own. At a minimum, I'd like a component to be able to report:

| Status | Description |
|-------|-------------|
| StatusOK | The component is operating as expected |
| StatusError | The component has encountered an "unexpected" error, look at the error for details |
| StatusStartupError | The component failed to start |
| StatusRecoverableError | The component is experiencing a likely transient error |
| StatusPermanentError | The component has encountered an error from which it will not recover; the component is giving up and will stop itself |
| StatusFatalError | The error from this component should trigger a shutdown of the entire collector |